src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -812,6 +816,646 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
   return new LateInlineVectorReboxingCallGenerator(method, inline_cg);
 }
 
+// Inline ScopedValue.get() call, pattern match the resulting subgraph, transform the subgraph to make it more amenable
+// to optimizations.
+class LateInlineScopedValueCallGenerator : public LateInlineCallGenerator {
+ private:
+  bool _process_result;
+  Node* _scoped_value_object;
+
+  class ScopedValueGetPatternMatcher : public StackObj {
+   private:
+    GraphKit& _kit;
+    Node* _scoped_value_object;
+    CallNode* _scoped_value_cache; // call to Thread.scopedValueCache()
+    IfNode* _cache_not_null_iff; // test that scopedValueCache() is not null
+    IfNode* _first_cache_probe_iff; // test for a hit in the cache with first hash
+    IfNode* _second_cache_probe_iff; // test for a hit in the cache with second hash
+    Node* _first_index_in_cache; // index in the cache for first hash
+    Node* _second_index_in_cache; // index in the cache for second hash
+    CallStaticJavaNode* _slow_call; // slowGet() call if any
+
+    bool match_cache_null_check_with_input(Node* maybe_cache, Node* maybe_nullptr, IfNode* iff) {
+      if (!maybe_cache->is_Proj() ||
+          !maybe_cache->in(0)->is_Call() ||
+          maybe_cache->in(0)->as_CallJava()->method()->intrinsic_id() != vmIntrinsics::_scopedValueCache) {
+        return false;
+      }
+      assert(maybe_nullptr->bottom_type() == TypePtr::NULL_PTR, "should be a test with null");
+      assert(_cache_not_null_iff == nullptr, "should only find one get_cache_if");
+      _cache_not_null_iff = iff;
+      assert(_scoped_value_cache == nullptr || _scoped_value_cache == maybe_cache->in(0),
+             "should only find one scoped_value_cache");
+      _scoped_value_cache = maybe_cache->in(0)->as_Call();
+      return true;
+    }
+
+    // Pattern matches:
+    // if ((objects = scopedValueCache()) != null) {
```

Comment:
```suggestion
    // if (scopedValueCache() != null) {
```
You don't use `objects` here, so it just confused me.

Reply:
I use snippets from the java code for `ScopedValue.get()` in the comments so it's easier to see what's being pattern matched.